src/pkcs15init/pkcs15-lib.c

```diff
@@ -1664,6 +1664,7 @@ sc_pkcs15init_generate_key(struct sc_pkcs15_card *p15card, struct sc_profile *pr
 
 err:
 	sc_pkcs15_free_object(object);
+	free(pubkey->alg_id);
```

Comment:
This is not what I meant -- now you are freeing it even if it was not copied, which might cause double frees.

I had in my mind modifying the block from line 1629 to happen in any case to make sure we do not modify the existing sc_algorith_id structure pointed by `pubkey_args` if it is provided by caller and not share it with the newly created `pubkey` object to avoid the double-free.

Reply:
My bad. You are right!